core/sdk/src/main/java/com/mongodb/stitch/core/services/internal/CoreStitchServiceClientImpl.java

```diff
@@ -195,14 +202,44 @@ public CoreStitchServiceClient withCodecRegistry(final CodecRegistry codecRegist
 
   @Override
   public void bind(final StitchServiceBinder binder) {
-    this.serviceBinders.add(new WeakReference<>(binder));
+    this.serviceBinders.put(new WeakReference<>(binder), Boolean.TRUE);
   }
 
   @Override
   public void onRebindEvent(final RebindEvent rebindEvent) {
-    final Iterator<WeakReference<StitchServiceBinder>> iterator = this.serviceBinders.iterator();
-    while (iterator.hasNext()) {
-      final WeakReference<StitchServiceBinder> weakReference = iterator.next();
+    boolean isActiveUserChange = false;
+    switch (rebindEvent.getType()) {
+      case AUTH_EVENT:
+        isActiveUserChange = ((AuthEvent) rebindEvent).getAuthEventType() ==
+            AuthEvent.Type.ACTIVE_USER_CHANGED;
+        break;
+      default:
+        break;
+    }
+    if (isActiveUserChange) {
+      final Iterator<WeakReference<Stream<?>>> streamIterator =
```

Comment:
[style nit] just for the sake of not doing too many things in one method, I would put the stuff in this conditional in a method called `closeAllocatedStreams()` and just call it here

Reply:
fair enough